pkg/storage/segments/repairer.go

```diff
@@ -186,57 +186,61 @@ func (repairer *Repairer) Repair(ctx context.Context, path storj.Path) (err erro
 		return Error.Wrap(err)
 	}
 
-	// Add the successfully uploaded pieces to the healthyPieces
+	// Add the successfully uploaded pieces to repairedPieces
+	var repairedPieces []*pb.RemotePiece
+	repairedMap := make(map[int32]bool)
 	for i, node := range successfulNodes {
 		if node == nil {
 			continue
 		}
-		healthyPieces = append(healthyPieces, &pb.RemotePiece{
+		piece := pb.RemotePiece{
 			PieceNum: int32(i),
 			NodeId:   node.Id,
 			Hash:     hashes[i],
-		})
-		healthyMap[int32(i)] = true
+		}
+		repairedPieces = append(repairedPieces, &piece)
+		repairedMap[int32(i)] = true
 	}
 
-	healthyLength := int32(len(healthyPieces))
+	healthyAfterRepair := int32(len(healthyPieces) + len(repairedPieces))
 	switch {
-	case healthyLength <= pointer.Remote.Redundancy.RepairThreshold:
+	case healthyAfterRepair <= pointer.Remote.Redundancy.RepairThreshold:
 		mon.Meter("repair_failed").Mark(1)
-	case healthyLength < pointer.Remote.Redundancy.SuccessThreshold:
+	case healthyAfterRepair < pointer.Remote.Redundancy.SuccessThreshold:
 		mon.Meter("repair_partial").Mark(1)
 	default:
 		mon.Meter("repair_success").Mark(1)
 	}
 
 	healthyRatioAfterRepair := 0.0
 	if pointer.Remote.Redundancy.Total != 0 {
-		healthyRatioAfterRepair = float64(healthyLength) / float64(pointer.Remote.Redundancy.Total)
+		healthyRatioAfterRepair = float64(healthyAfterRepair) / float64(pointer.Remote.Redundancy.Total)
 	}
 	mon.FloatVal("healthy_ratio_after_repair").Observe(healthyRatioAfterRepair)
 
-	// if partial repair, include "unhealthy" pieces that are not duplicates
-	revisedPointerPieces := healthyPieces
-	if healthyLength < pointer.Remote.Redundancy.SuccessThreshold {
-		for _, p := range unhealthyPieces {
-			if _, ok := healthyMap[p.GetPieceNum()]; !ok {
-				revisedPointerPieces = append(revisedPointerPieces, p)
+	// if partial repair, leave "unhealthy" pieces in the pointer that were not repaired
+	unhealthyPiecesToRemove := unhealthyPieces
+	if healthyAfterRepair < pointer.Remote.Redundancy.SuccessThreshold {
+		for _, p := range unhealthyPiecesToRemove {
+			num := p.GetPieceNum()
+			if !repairedMap[num] {
+				// leave only repaired pieces in the slice, unrepaired
+				// unhealthy pieces are not removed from the pointer
+				unhealthyPiecesToRemove = append(unhealthyPiecesToRemove[:num], unhealthyPiecesToRemove[num+1:]...)
```

Comment:
Oh, you are right! We should use the loop index when removing the item from `unhealthyPiecesToRemove`. Fixed.

Reply:
Well  I am still not sure if that is sufficient. Suppose you have a `list` of 10 elements, and you are iterating over it with 
`for i, item := range list {`
If you are iterating and you are on index 4, and you say `list = append(list[:i], list[i+1:])`, what do you expect in the next iteration? The next item we *want* is at index 4 now (because we removed the current item). But will the `for` loop start the next iteration with `i=4` again, or will it continue to `i=5` and skip an item?

Reply:
see this: https://play.golang.org/p/jFZJRWoaZp5

Reply:
omg, I completely screwed it. Let me think about a better solution.

Reply:
I think we could just append the pieces to a new list and keep the one we are iterating over static. Then, after the loop, we can set the one we looped over to the new one. This is similar to what was happening before

Reply:
I think I fixed it now, following your suggestion.